Adding xpack code for ES cluster stats metricset

ycombinator opened the pull request:
This PR teaches the `elasticsearch/cluster_stats` metricset to query the appropriate Elasticsearch HTTP APIs and index `cluster_stats` documents into `.monitoring-es-6-mb-*` indices. These documents should be compatible in structure to `cluster_stats` documents in the current `.monitoring-es-6-*` indices indexed via the internal monitoring method.

ruflin commented on metricbeat/module/elasticsearch/cluster_stats/data_xpack.go:
```diff
@@ -0,0 +1,238 @@
+		return err
+	}
+
+	dataMS := common.MapStr(data)
```
Perhaps add a type assertion even though this should never fail I think.

ruflin replied:
Ignore my comment, above is good.

ruflin commented on metricbeat/module/elasticsearch/cluster_stats/data_xpack.go:
```diff
@@ -0,0 +1,238 @@
+	passthruFields := []string{
+		"indices.segments.file_sizes", // object with dynamic keys
+		"nodes.versions", // array of strings
+		"nodes.os.names", // array of objects
```
Is it a list of names : `["a", "b"]` or nested objects `[{"a":"b"}, {"c":"d"}]`. Just curious.

ycombinator replied:
It is a list of objects, for example:

```json
[
                  {
                    "name": "Mac OS X",
                    "count": 1
                  }
]
```

ruflin replied:
:-(

ruflin commented on metricbeat/module/elasticsearch/cluster_stats/data_xpack.go:
```diff
@@ -0,0 +1,238 @@
+
+	dataMS := common.MapStr(data)
+
+	passthruFields := []string{
```
In general it seems these are values which outside x-pack reporting should be reported by the specific metricsets and to get all names for example it would be an aggregation based on the cluster id?

ruflin commented on metricbeat/module/elasticsearch/elasticsearch.go:
```diff
@@ -167,3 +170,39 @@ func GetNodeInfo(http *helper.HTTP, uri string, nodeID string) (*NodeInfo, error
 }
+
+// GetLicense returns license information
+func GetLicense(http *helper.HTTP, resetURI string) (map[string]interface{}, error) {
```
Should we cache this value for a few minute as the license is hopefully very rarely going to change?

ycombinator replied:
Good idea. Will do.

ycombinator replied:
Implemented in e72f9c90c. Curious to hear your thoughts on the implementation — too complex, too simple, just right for now?

ruflin commented on metricbeat/module/elasticsearch/cluster_stats/data_xpack.go:
```diff
@@ -0,0 +1,238 @@
+		"timestamp":     common.Time(time.Now()),
+		"interval_ms":   m.Module().Config().Period / time.Millisecond,
+		"type":          "cluster_stats",
+		"license":       license,
```
Is the full license response reported here?

ycombinator replied:
Yes, it is the response from the `GET _xpack/license` Elasticsearch API, which looks like this:

```js
{
  "license": {
    "status": "active",
    "uid": "f80c1fb5-75e9-4536-be2d-a768b02abb46",
    "type": "basic",
    "issue_date": "2018-08-06T23:47:09.619Z",
    "issue_date_in_millis": 1533599229619,
    "max_nodes": 1000,
    "issued_to": "elasticsearch",
    "issuer": "elasticsearch",
    "start_date_in_millis": -1
  }
}
```

ycombinator commented on metricbeat/module/elasticsearch/cluster_stats/data_xpack.go:
```diff
@@ -0,0 +1,238 @@
+		"type":          "cluster_stats",
+		"license":       license,
+		"version":       info.Version.Number,
+		"cluster_stats": clusterStats,
```
@ruflin As you can see here, the document that is indexed into monitoring contains not only `clusterStats` but also `clusterState` (an abridged version of it) and `stackStats`.

For [`clusterState`](https://github.com/elastic/beats/pull/7810/files#diff-5c3e5a2da8527b611dff0cc47e744173R202) and [`stackStats`](https://github.com/elastic/beats/pull/7810/files#diff-5c3e5a2da8527b611dff0cc47e744173R213) we are simply calling the corresponding Elasticsearch APIs, doing a tiny bit of massaging of the response data, and then passing the resulting structure through over here. We are definitely not doing a complete schema conversion like we are doing for [`clusterStats`](https://github.com/elastic/beats/pull/7810/files#diff-5c3e5a2da8527b611dff0cc47e744173R34).

It makes me wonder: should we treat `clusterStats` special over here? Or conversely, should we do complete schema conversions for `clusterState` and `stackStats` too?

ruflin replied:
How is the internal reporting in ES happening? Is it just taking all the existing fields? If yes then we should do the same here.

ycombinator replied:
Yes, it just takes all the existing fields and passes them through. Here is the relevant code:

https://github.com/elastic/elasticsearch/blob/b87f3062b77cab7888e9037b4996f2c26db12816/x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/collector/cluster/ClusterStatsMonitoringDoc.java#L141-L147

ycombinator commented on metricbeat/module/elasticsearch/cluster_stats/data_xpack.go:
```diff
@@ -0,0 +1,339 @@
+}
+
+// computeNodesHash computes a simple hash value that can be used to determine if the nodes listing has changed since the last report.
+func computeNodesHash(clusterState common.MapStr) (int32, error) {
```
@pickypg If possible I'd like you to review this function. It attempts to port over the logic implemented in https://github.com/elastic/elasticsearch/blob/93ada98eb54d99b9d98ed51778bd6422d18fd392/x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/collector/cluster/ClusterStatsMonitoringDoc.java#L180-L195.

ycombinator commented:
@ruflin This PR is (finally!) ready for review now. Thanks.

ruflin commented on metricbeat/module/elasticsearch/cluster_stats/data_xpack.go:
```diff
@@ -0,0 +1,370 @@
+	for _, value := range nodes {
+		nodeData, ok := value.(map[string]interface{})
+		if !ok {
+			return 0, fmt.Errorf("Node data is not a map")
```
Nit: In general errors which are returned and not printed should start lower case as they could be embeded in other errors. I wonder why hound did not complain. Same applies to most other errors in this PR.

ruflin commented on metricbeat/module/elasticsearch/elasticsearch.go:
```diff
@@ -21,17 +21,54 @@ import (
 var clusterIDCache = map[string]string{}
 
+// Global cache for license information. Assumption is that license information changes infrequently
+type _licenseCache struct {
```
Why do you start with a _ here? I would prefer not to have _ in var names.

ycombinator replied:
I don't like the `_` prefix here either so I'm open to other suggestions. The reason I did this was because, just a few lines below, I want to create the actual singleton cache instance variable. I wanted _that_ variable to have the "real" name, `licenseCache`, so I had to come up with some other name for the type.

ruflin commented on metricbeat/module/elasticsearch/elasticsearch.go:
```diff
@@ -21,17 +21,54 @@ import (
+
+var licenseCache = &_licenseCache{}
+
+func (c *_licenseCache) get() common.MapStr {
```
nit: I would put the private and cache related code to the bottom of the file as I would assume most people come to the file to look at the public methods.

ruflin commented on metricbeat/module/elasticsearch/elasticsearch.go:
```diff
@@ -167,3 +204,51 @@ func GetNodeInfo(http *helper.HTTP, uri string, nodeID string) (*NodeInfo, error
 	return nil, fmt.Errorf("no node matched id %s", nodeID)
 }
+
+// GetLicense returns license information
```
Could you add a note here that the license is cached?

ycombinator commented:
@ruflin This is ready for another round of review. I've addressed most of your feedback; only this thread remains unresolved: https://github.com/elastic/beats/pull/7810#discussion_r210520948

ycombinator commented:
> Will this need to be backported to 6.x?
Yes, just added the necessary labels and put up the backport PR here: https://github.com/elastic/beats/pull/8000
